src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1699,6 +1703,33 @@ impl Suggestion {
             Self::RemoveDisambiguator => path_str.into(),
         }
     }
+
+    fn as_help_span(
+        &self,
+        path_str: &str,
+        ori_link: &str,
+        sp: rustc_span::Span,
+    ) -> (String, rustc_span::Span) {
+        // FIXME: if this is an implied shortcut link, it's bad style to suggest `@`
+        match self {
+            Self::Prefix(prefix) => {
+                let sp = match ori_link.find('@') {
+                    Some(index) => sp.with_hi(sp.lo() + BytePos(index as u32 + 1)),
+                    None => sp.shrink_to_lo(),
+                };
+                (format!("{}@", prefix), sp)
+            }
+            Self::Function => ("()".to_string(), sp.shrink_to_hi()),
+            Self::Macro => {
+                let sp = match ori_link.find('(') {
+                    Some(index) => sp.with_hi(sp.lo() + BytePos(index as _)),
+                    None => sp,
+                };
+                ("!".to_string(), sp.shrink_to_hi())
+            }
+            Self::RemoveDisambiguator => (path_str.into(), sp),
+        }
+    }
```

Comment:
I _think_ that the following code would do the trick. You will have to `use` `Span` (you already have a use list from `rustc_span`) and to change the caller to take a `Vec` instead of the tuple. You can feed that straight to `multpart_suggestion_verbose` and unless I made a mistake (it's been known to happen), it should be given correct suggestions now.

```suggestion
    fn as_help_span(
        &self,
        path_str: &str,
        ori_link: &str,
        sp: Span,
    ) -> Vec<(String, Span)> {
        // FIXME: if this is an implied shortcut link, it's bad style to suggest `@`
        let inner_sp = match ori_link.find("()") {
            Some(index) => sp.with_hi(sp.lo() + BytePos(index as _)),
            None => sp,
        };
        let inner_sp = match ori_link.find('@') {
            Some(index) => inner_sp.with_lo(inner_sp.lo() + BytePos(index as u32 + 1)),
            None => sp,
        };
        match self {
            Self::Prefix(prefix) => {
                let mut sugg = vec![(format!("{}@", prefix), sp.with_hi(inner_sp.lo()))];
                if sp.hi() != inner_sp.hi() {
                    sugg.push(("".to_string(), inner_sp.shrink_to_hi().with_hi(sp.hi()));
                }
                sugg
            }
            Self::Function => {
                let mut sugg = vec![("()".to_string(), inner_sp.shrink_to_hi().with_hi(sp.hi()))];
                if sp.lo() != inner_sp.lo() {
                    sugg.push(("".to_string(), inner_sp.shrink_to_lo().with_lo(sp.lo()));
                }
                sugg
            }
            Self::Macro => {
                let mut sugg = vec![("!".to_string(), inner_sp.shrink_to_hi().with_hi(sp.hi()))];
                if sp.lo() != inner_sp.lo() {
                    sugg.push(("".to_string(), inner_sp.shrink_to_lo().with_lo(sp.lo()));
                }
                sugg
            }
            Self::RemoveDisambiguator => {
                let mut sugg = vec![];
                if sp.lo() != inner_sp.lo() {
                    sugg.push(("".to_string(), sp.with_hi(inner_sp.lo())));
                }
                if sp.hi() != inner_sp.hi() {
                    sugg.push(("".to_string(), inner_sp.with_hi(sp.hi())));
                }
                sugg
            }
        }
```

Reply:
Wo thanks! A few things to modify but this completely fixed the issues I had with using `multipart_suggestion`!